src/datadog/grok/matchers/date.rs

```diff
@@ -241,120 +243,141 @@ pub fn time_format_to_regex(format: &str, with_captures: bool) -> Result<RegexRe
 }
 
 pub fn apply_date_filter(value: &Value, filter: &DateFilter) -> Result<Value, GrokRuntimeError> {
-    let timestamp = match value {
-        Value::Bytes(bytes) => {
-            let mut value = String::from_utf8_lossy(bytes).into_owned();
-            // Ideally this Z should be quoted in the pattern, but DataDog supports this as a special case:
-            // yyyy-MM-dd'T'HH:mm:ss.SSSZ - e.g. 2016-09-02T15:02:29.648Z
-            if value.ends_with('Z') && filter.original_format.ends_with('Z') {
-                value.pop(); // drop Z
-                value.push_str("+0000");
-            }
-            if let Some(tz) = filter
-                .regex
-                .captures(&value)
-                .and_then(|caps| caps.name("tz"))
-            {
-                let tz = tz.as_str();
-                let tz: Tz = tz.parse().map_err(|error| {
+    let original_value = String::from_utf8_lossy(value.as_bytes().ok_or_else(|| {
+        GrokRuntimeError::FailedToApplyFilter(filter.to_string(), value.to_string())
+    })?);
+    let (strp_format, mut datetime) =
+        adjust_strp_format_and_value(&filter.strp_format, &original_value);
+
+    // Ideally this Z should be quoted in the pattern, but DataDog supports this as a special case:
+    // yyyy-MM-dd'T'HH:mm:ss.SSSZ - e.g. 2016-09-02T15:02:29.648Z
+    if datetime.ends_with('Z') && filter.original_format.ends_with('Z') {
+        datetime.pop(); // drop Z
+        datetime.push_str("+0000");
+    };
+
+    if let Some(tz) = filter
+        .regex
+        .captures(&original_value)
+        .and_then(|caps| caps.name("tz"))
+    {
+        let tz = tz.as_str();
+        let tz: Tz = tz.parse().map_err(|error| {
+            warn!(message = "Error parsing tz", tz = %tz, % error);
+            GrokRuntimeError::FailedToApplyFilter(filter.to_string(), original_value.to_string())
+        })?;
+        replace_sec_fraction_with_dot(filter, &mut datetime);
+        let naive_date = NaiveDateTime::parse_from_str(&datetime, &strp_format).map_err(|error|
+        {
+            warn!(message = "Error parsing date", value = %original_value, format = %strp_format, % error);
+            GrokRuntimeError::FailedToApplyFilter(
+                filter.to_string(),
+                original_value.to_string(),
+            )
+        })?;
+        let dt = tz
+            .from_local_datetime(&naive_date)
+            .single()
+            .ok_or_else(|| {
+                GrokRuntimeError::FailedToApplyFilter(
+                    filter.to_string(),
+                    original_value.to_string(),
+                )
+            })?;
+        Ok(Value::from(
+            Utc.from_utc_datetime(&dt.naive_utc()).timestamp_millis(),
+        ))
+    } else {
```

Comment:
This is a totally non-blocking comment but an observation- there are a ton of nested if/else, if let  branches in this function and that makes it pretty hard to follow the logic when it gets so large. It'd be helpful to have either helper functions or a comment somewhere that describes the behavior.

Reply:
Totally agree - I added a [task](https://datadoghq.atlassian.net/browse/OPA-1935) to address this issue.